src/sources/util/finalizer.rs

```diff
@@ -0,0 +1,91 @@
+use crate::event::BatchStatusReceiver;
+use crate::shutdown::ShutdownSignal;
+use futures::{future::Shared, stream::FuturesOrdered, FutureExt, StreamExt};
+use std::future::Future;
+use std::pin::Pin;
+use std::task::Poll;
+use tokio::sync::mpsc;
+
+/// The `OrderedFinalizer` framework here is a mechanism for marking
+/// events from a source as done in a single background task *in the
+/// order they are received from the source*. The type `T` is the
+/// source-specific data associated with each entry to be used to
+/// complete the finalization.
+pub(crate) struct OrderedFinalizer<T> {
+    sender: Option<mpsc::UnboundedSender<(BatchStatusReceiver, T)>>,
+}
+
+// TODO: Rewrite `apply_done` below into a trait once there are more
```

Comment:
Rewrite how?

Reply:
I was thinking to make `OrderedFinalizer` generic on a new trait that has an `apply_done` method. The original framework had the possibility for calling `apply_done` on more than one message at once, which would then require start and end methods. I'll just drop this TODO note for now.

Reply:
Makes sense. Thanks.